pkg/server/tenant_migration.go

```diff
@@ -0,0 +1,173 @@
+			"binary's minimum supported version %s for SQL server instance %d",
+			targetCV, tenantVersion.BinaryMinSupportedVersion(),
+			instanceID)
+		return err
+	}
+
+	if tenantVersion.BinaryVersion().Less(targetCV.Version) {
+		err := errors.Newf("sql server %s is running a binary version %s which is "+
+			"less than the attempted upgrade version %s",
+			instanceID.String(),
+			tenantVersion.BinaryVersion(), targetCV)
+		return errors.WithHintf(err,
+			"upgrade sql server %d binary to version %s (or higher) to allow tenant upgrade to succeed",
+			instanceID,
+			targetCV,
+		)
+	}
+	return nil
+}
+
+// ValidateTargetClusterVersion implements the MigrationServer interface.
+// It's used to verify that each tenant server is running a binary that's recent
+// enough to support the upgrade.
+func (m *TenantMigrationServer) ValidateTargetClusterVersion(
+	ctx context.Context, req *serverpb.ValidateTargetClusterVersionRequest,
+) (*serverpb.ValidateTargetClusterVersionResponse, error) {
+	ctx = m.sqlServer.AnnotateCtx(ctx)
+	ctx = logtags.AddTag(ctx, "validate-tenant-cluster-version", nil)
+
+	if err := validateTargetClusterVersion(ctx,
+		m.sqlServer.settingsWatcher.GetTenantClusterVersion(),
+		req.ClusterVersion,
+		m.sqlServer.SQLInstanceID()); err != nil {
+		return nil, err
+	}
+
+	resp := &serverpb.ValidateTargetClusterVersionResponse{}
+	return resp, nil
+}
+
+// BumpClusterVersion implements the MigrationServer interface. It's used to
+// inform us of a cluster version bump. When called for tenants we update the
+// tenant server's active cluster version.
+func (m *TenantMigrationServer) BumpClusterVersion(
+	ctx context.Context, req *serverpb.BumpClusterVersionRequest,
+) (*serverpb.BumpClusterVersionResponse, error) {
+	opName := upgradecluster.BumpClusterVersionOpName
+	ctx = m.sqlServer.AnnotateCtx(ctx)
+	ctx = logtags.AddTag(ctx, opName, nil)
+
+	if err := m.sqlServer.stopper.RunTaskWithErr(ctx, opName, func(
+		ctx context.Context,
+	) error {
+		m.Lock()
+		defer m.Unlock()
+		return bumpTenantClusterVersion(ctx,
+			m.sqlServer.settingsWatcher.GetTenantClusterVersion(),
+			*req.ClusterVersion,
+			m.sqlServer.SQLInstanceID())
+	}); err != nil {
+		return nil, err
+	}
+	return &serverpb.BumpClusterVersionResponse{}, nil
+}
+
+// bumpTenantClusterVersion increases the active version for a tenant server.
+// This logic is much more straightforward for tenant upgrades because tenants
+// don't store an engine version in addition to the active version.
+func bumpTenantClusterVersion(
+	ctx context.Context,
+	tenantCV clusterversion.Handle,
+	newCV clusterversion.ClusterVersion,
+	instanceID base.SQLInstanceID,
+) error {
+	log.Infof(ctx, "bumping cluster version from %v to %v on instance %s", tenantCV, newCV, instanceID.String())
+	activeCV := tenantCV.ActiveVersion(ctx)
+	if !activeCV.Less(newCV.Version) {
+		// Nothing to do.
+		return nil
+	}
+
+	// It's possible that between the time we performed the first validation of
```

Comment:
I find this comment hard to follow. You need to have a lot of context paged in for it to make sense. Only after coming back to it did I follow. The "as early as possible" still isn't very obvious to me. Can we iterate on this comment so its a bit more free-standing?